extensions/amp-story/1.0/amp-story.js

```diff
@@ -2375,6 +2360,21 @@ export class AmpStory extends AMP.BaseElement {
       nextPageEl.setAttribute(Attributes.RETURN_TO, pageToBeInsertedId);
     }
 
+    // Adjust the ad page's position in this.adPages_ array to reflect the actual.
+    const adPageIndex = this.getPageIndexById(pageToBeInsertedId);
+    if (adPageIndex != -1) {
+      this.pages_.splice(adPageIndex, 1);
+    }
+    this.pages_.splice(
+      this.getPageIndexById(pageBeforeId) + 1,
+      0,
+      pageToBeInserted
+    );
+
+    this.storeService_.dispatch(
```

Comment:
From this reordering, why not insert the ad in the right place from the start, if possible?

Reply:
Right this is so that we can create/preload the ad element before we know if it will actually load, or even if the server will respond with an ad.